src/backend/lib/upgrade-download.js

```diff
@@ -0,0 +1,303 @@
+                    searchLog("skipped upgrade candidate", upgrade);
+                  }
+                });
+              });
+            } catch (err) {
+              searchLog("JSON parse error:", err);
+            } finally {
+              release();
+            }
+          });
+        })
+        .once("error", err => {
+          searchLog("HTTP error:", err);
+          release();
+        });
+    });
+  }
+
+  stop() {
+    searchLog("stopping..");
+    this.stateLock.writeLock(release => {
+      if (this.state !== UpgradeState.Search.Searching) return release();
+
+      this.discovery.destroy(err => {
+        if (err) {
+          this.setState(UpgradeState.Search.Error, err);
+          searchLog("..stopped, with error:", err);
+        } else {
+          this.setState(UpgradeState.Search.Idle, null);
+          searchLog("..stopped");
+        }
+        release();
+      });
+    });
+  }
+
+  setState(state, context) {
+    this.state = state;
+    this.context = context;
+    this.emit("state", state, context);
+  }
+
+  // UpgradeOption -> Bool
+  isUpgradeCandidate(upgrade) {
+    if (upgrade.arch.indexOf(this.storage.getLocalArch()) === -1) return false;
+    if (upgrade.platform !== this.storage.getLocalPlatform()) return false;
+    if (!semver.valid(upgrade.version)) return false;
+    if (!semver.gt(upgrade.version, this.storage.getLocalVersion()))
+      return false;
+    if (this.context.upgrades.some(u => u.hash === upgrade.hash)) return false;
+    return true;
+  }
+}
+
+// Responsible for downloading a given UpgradeOption and tracking its download
+// progress.
+class Download extends EventEmitter {
+  constructor(storage) {
+    super();
+    this.state = null;
+    this.context = null;
+    this.setState(UpgradeState.Download.Idle, null);
+    this.storage = storage;
+  }
+
+  setState(state, context) {
+    this.state = state;
+    this.context = context;
+    this.emit("state", state, context);
+  }
+
+  download(option) {
+    if (this.state !== UpgradeState.Download.Idle) return;
+    this.setState(UpgradeState.Download.Downloading, {
+      sofar: 0,
+      total: option.size,
+    });
+    downloadLog("starting download of upgrade", option);
+
+    const url = `/content/${option.id}`;
+    http
+      .get({ hostname: option.host, port: option.port, path: url }, res => {
+        const filename = option.hash;
+        let sofar = 0;
+        const progress = through((chunk, enc, next) => {
+          sofar += chunk.length;
+          this.setState(UpgradeState.Download.Downloading, {
+            sofar,
+            total: option.size,
+          });
+          next(null, chunk);
+        });
+        const ws = this.storage.createApkWriteStream(
+          filename,
+          option.version,
+          option.hash,
+          err => {
+            if (err) this.setState(UpgradeState.Download.Error, err);
+            else this.setState(UpgradeState.Download.Downloaded, null);
+          }
```

Comment:
It doesn't seem to make sense to mix a stream and a callback, because we end up with two ways this completes: when the `pump` pipeline completes, and then when this callback is called, which happen at different times. I feel like `createApkWriteStream` should return a writable stream which will `finish` when the APK is written.